### PR TITLE
net: shell: UDP commands

### PR DESCRIPTION
**Pull request by hubertmis**
This patch adds to network shell set of basic commands for UDP protocol
to receive and send datagrams. It is compatible only with IPv6. It
can be extended later with IPv4 support as well.

Signed-off-by: Hubert Miś <hubert.mis@nordicsemi.no>

### CODE REVIEW

**Review by jukkar — CHANGES_REQUESTED** [review]

File: subsys/net/ip/net_shell.c
Comment (hubertmis) [reply subsys/net/ip/net_shell.c:4813]: I'm testing it with OpenThread, I don't have any IPv4 capable applications to verify IPv4 variant. I would prefer to limit scope of this PR to IPv6 and if someone would like to use it with IPv4, it should be straightforward to extend.

File: subsys/net/ip/net_shell.c
Comment (hubertmis) [reply subsys/net/ip/net_shell.c:4803]: I got inspiration from TCP functions. It seems the limitation you described here is still present in `tcp_connect()`. Or am I missing something?

Comment (jukkar) [reply subsys/net/ip/net_shell.c:4803]: Yes, the same limitation is in TCP too and should be fixed at some point.

File: subsys/net/ip/net_shell.c
Comment (jukkar) [reply subsys/net/ip/net_shell.c:4813]: > I'm testing it with OpenThread,

Ok, that is good. But this is trivial to test using Qemu which by default provides dual IPv6 and IPv4 support.

**Review by rlubos — COMMENTED** [review]
Looks good, but I agree with Jukka that it should be pretty straightforward to test IPv4 with QEMU so it shouldn't be a big deal to add/test IPv4 as well for completeness.

I also wonder, while current solution is consistent with the TCP shell module implementation, we should likely be moving towards socket API usage in modules like this? There's an open issue to convert TCP shell to socket API. But I don't think it should be a blocker, perhaps both TCP and UDP could be reimplemented together at some point in the future.

**Comment by hubertmis** [comment]
> that it should be pretty straightforward to test IPv4 with QEMU so it shouldn't be a big deal to add/test IPv4 as well for completeness

I created this patch while I was testing my OpenThread project. I'm happy to share my work with community and add UDP commands for IPv6. I don't think I'll have enough free time soon to setup QEMU and add IPv4 support. But if you think missing IPv4 is a blocker, we can close this PR for now.

I'm also OK if someone modified this PR and add IPv4 support before it is merged.

**Comment by jukkar** [comment]
See #32750 I added IPv4 support. For some reason I could not push to this PR so created a new one.

**Comment by jukkar** [comment]
> perhaps both TCP and UDP could be reimplemented together at some point in the future.

Ack to this, we should convert both the TCP and UDP using sockets but probably not in this PR

**Comment by hubertmis** [comment]
> For some reason I could not push to this PR so created a new one.

Interesting. I verified, the box allowing edits by maintainers is checked. Anyway, I reviewed #32750. It looks nearly ready to be merged.

**Comment by hubertmis** [comment]
Suppressed by #32750

**Pull request closed without merging**
